circuit-codegen/src/main/kotlin/com/slack/circuit/codegen/CircuitSymbolProcessorProvider.kt

```diff
@@ -179,59 +231,149 @@ private class CircuitSymbolProcessor(
+              addProperty(
+                PropertySpec.builder(param.name, param.type, KModifier.PRIVATE)
+                  .initializer(param.name)
+                  .build()
+              )
+            }
+          }
+        }
+    val screenBranch =
+      if (screenIsObject) {
+        CodeBlock.of("%T", screenType)
+      } else {
+        CodeBlock.of("is·%T", screenType)
+      }
     val typeSpec =
       when (factoryType) {
         FactoryType.PRESENTER ->
-          buildPresenterFactory(className, annotatedElement, screenType, codeBlock)
-        FactoryType.UI -> buildUiFactory(className, annotatedElement, screenType, codeBlock)
-      }
-
-    val finalSpec =
-      extensions.fold(typeSpec) { spec, extension ->
-        extension.process(spec, annotatedElement, factoryType)
+          builder.buildPresenterFactory(annotatedElement, screenBranch, codeBlock)
+        FactoryType.UI -> builder.buildUiFactory(annotatedElement, screenBranch, codeBlock)
       }
 
-    FileSpec.get(packageName, finalSpec).writeTo(codeGenerator = codeGenerator, aggregating = false)
+    FileSpec.get(packageName, typeSpec).writeTo(codeGenerator = codeGenerator, aggregating = false)
   }
 }
 
-private fun buildUiFactory(
-  className: String,
+private data class AssistedType(
+  val factoryName: String,
+  val type: TypeName,
+  val name: String,
+)
+
+private fun KSFunctionDeclaration.assistedParameters(
+  symbols: CircuitSymbols,
+  logger: KSPLogger,
+  screenType: KSType,
+  allowNavigator: Boolean
+): CodeBlock {
+  return buildSet {
+      for (param in parameters) {
+        val type = param.type.resolve()
+        if (symbols.screen.isAssignableFrom(type)) {
+          if (screenType.isAssignableFrom(type)) {
+            val added = add(AssistedType("screen", type.toTypeName(), param.name!!.getShortName()))
+            if (!added) {
+              logger.error("Multiple parameters of type $type are not allowed.", param)
+            }
+          } else {
+            logger.error("Screen type mismatch. Expected $screenType but found $type", param)
+          }
+        } else if (symbols.navigator.isAssignableFrom(type)) {
+          if (allowNavigator) {
+            val added =
+              add(AssistedType("navigator", type.toTypeName(), param.name!!.getShortName()))
+            if (!added) {
+              logger.error("Multiple parameters of type $type are not allowed.", param)
+            }
+          } else {
+            logger.error(
+              "Navigator type mismatch. Navigators are not injectable on this type.",
+              param
+            )
+          }
+        } else if (symbols.circuitConfig.isAssignableFrom(type)) {
+          val added =
+            add(AssistedType("circuitConfig", type.toTypeName(), param.name!!.getShortName()))
+          if (!added) {
+            logger.error("Multiple parameters of type $type are not allowed.", param)
+          }
+        }
```

Comment:
nit: there's some duplication here. You could pull some of it out and improve readability
```
private fun MutableSet<AssistedType>.add(type: KSType, param: KSValueParameter, logger: KSPLogger, assistedType: AssistedType) {
  val added = add(assistedType)
  if (!added) {
    logger.error("Multiple parameters of type $type are not allowed.", param)
  }
}

...

if (screenType.isAssignableFrom(type)) {
  add(type, param, logger, AssistedType("screen", type.toTypeName(), param.name!!.getShortName()))
} else {
  logger.error("Screen type mismatch. Expected $screenType but found $type", param)
}

```